src/DMLQueryBuilder.php

```diff
@@ -140,89 +131,72 @@ public function upsert(
         }
 
         $on = $this->queryBuilder->buildCondition($onCondition, $params);
-        /** @psalm-var string[] $placeholders */
+
         [, $placeholders, $values, $params] = $this->prepareInsertValues($table, $insertColumns, $params);
 
         if (!empty($placeholders)) {
             $usingSelectValues = [];
-            /** @psalm-var string[] $insertNames */
+
             foreach ($insertNames as $index => $name) {
                 $usingSelectValues[$name] = new Expression($placeholders[$index]);
             }
 
-            /** @psalm-var array $params */
-            $usingValues = $this->queryBuilder->buildSelect($usingSelectValues, $params) . ' ' . $this->queryBuilder->buildFrom(['DUAL'], $params);
+            $values = $this->queryBuilder->buildSelect($usingSelectValues, $params)
+                . ' ' . $this->queryBuilder->buildFrom(['DUAL'], $params);
         }
 
         $insertValues = [];
-        $mergeSql = 'MERGE INTO '
-            . $this->quoter->quoteTableName($table)
-            . ' '
-            . 'USING (' . ($usingValues ?? ltrim((string) $values, ' '))
-            . ') "EXCLUDED" '
-            . "ON ($on)";
-
-        /** @psalm-var string[] $insertNames */
-        foreach ($insertNames as $name) {
-            $quotedName = $this->quoter->quoteColumnName($name);
-
-            if (strrpos($quotedName, '.') === false) {
-                $quotedName = '"EXCLUDED".' . $quotedName;
-            }
+        $mergeSql = 'MERGE INTO ' . $quotedTableName . ' USING (' . $values . ') "EXCLUDED" ON (' . $on . ')';
 
-            $insertValues[] = $quotedName;
+        foreach ($insertNames as $quotedName) {
+            $insertValues[] = '"EXCLUDED".' . $quotedName;
         }
 
         $insertSql = 'INSERT (' . implode(', ', $insertNames) . ')' . ' VALUES (' . implode(', ', $insertValues) . ')';
 
-        if ($updateColumns === false) {
+        if ($updateColumns === false || $updateNames === []) {
```

Comment:
```suggestion
        if (empty($updateColumns)) {
```

It's a little more productive.

Reply:
not correct suggestion